File: October/20/lrucache.cpp
// LeetCode 146: LRU Cache
// Explanation:
// 1. Use unordered_map for O(1) key lookup.
// 2. Use list for maintaining usage order (front = most recent).
// 3. Move used items to front; evict least recent when over capacity.
// Time Complexity: O(1)
// Space Complexity: O(capacity)

#include <unordered_map>
#include <list>
using namespace std;

class LRUCache {
    int cap;
    list<pair<int, int>> cache;
    unordered_map<int, list<pair<int, int>>::iterator> map;
public:
    LRUCache(int capacity) : cap(capacity) {}

    int get(int key) {
        if (map.find(key) == map.end()) return -1;
        auto it = map[key];
        int value = it->second;
        cache.erase(it);
        cache.push_front({key, value});
        map[key] = cache.begin();
        return value;
    }

    void put(int key, int value) {
        if (map.find(key) != map.end()) {
            cache.erase(map[key]);
        } else if (cache.size() == cap) {
            auto del = cache.back();
            map.erase(del.first);
            cache.pop_back();
        }
        cache.push_front({key, value});
        map[key] = cache.begin();
    }
};

#include <iostream>
int main() {
    LRUCache lru(2);
    lru.put(1, 1);
    lru.put(2, 2);
    cout << lru.get(1) << endl; // Output: 1
    lru.put(3, 3);
    cout << lru.get(2) << endl; // Output: -1
    lru.put(4, 4);
    cout << lru.get(1) << endl; // Output: -1
    cout << lru.get(3) << endl; // Output: 3
    cout << lru.get(4) << endl; // Output: 4
}

File: October/20/lrucache.py
# LeetCode 146: LRU Cache
# Explanation:
# 1. Implement a Least Recently Used (LRU) cache with O(1) time for get and put.
# 2. Use OrderedDict to maintain insertion order — most recently used items move to the end.
# 3. When capacity is exceeded, remove the least recently used item (first one).
# Time Complexity: O(1) for both get and put.
# Space Complexity: O(capacity)

from collections import OrderedDict

class LRUCache:
    def __init__(self, capacity: int):
        self.cache = OrderedDict()
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key)
        return self.cache[key]

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)
        self.cache[key] = value
        if len(self.cache) > self.capacity:
            self.cache.popitem(last=False)
